src/Commands/core/ArchiveDumpCommands.php

```diff
@@ -0,0 +1,585 @@
+                            $this->logger()->info(dt(
+                                'Path excluded (!exclude): !path',
+                                ['!exclude' => $exclude, '!path' => $localFileName]
+                            ));
+
+                            return false;
+                        }
+                    }
+
+                    $this->validateSensitiveData($file, $localFileName);
+
+                    return true;
+                }
+            )
+        );
+    }
+
+    /**
+     * Creates "database" archive component and returns the absolute path.
+     *
+     * @param array $options
+     *   The command options.
+     *
+     * @return string
+     *   The full path to the database archive component directory.
+     *
+     * @throws \Exception
+     *
+     * @see \Drush\Commands\sql\SqlCommands::dump()
+     */
+    private function getDatabaseComponentPath(array $options): string
+    {
+        $this->logger()->info(dt('Creating database SQL dump file...'));
+        $databaseArchiveDir = Path::join($this->archiveDir, self::COMPONENT_DATABASE);
+        $this->filesystem->mkdir($databaseArchiveDir);
+
+        $options['result-file'] = Path::join($databaseArchiveDir, self::SQL_DUMP_FILE_NAME);
+        $sql = SqlBase::create($options);
+        if (false === $sql->dump()) {
+            throw new Exception('Unable to dump database. Rerun with --debug to see any error message.');
+        }
+
+        return $databaseArchiveDir;
+    }
+
+    /**
+     * Returns the list of regular expressions to match paths.
+     *
+     * @param array $paths
+     *   The list of paths to match.
+     *
+     * @return array
+     */
+    private function getRegexpsForPaths(array $paths): array
+    {
+        return array_map(
+            fn($path) => sprintf('#^%s$#', addslashes(trim($path))),
+            $paths
+        );
+    }
+
+    /**
+     * Returns docroot directory name with trailing escaped slash for a "web" docroot site for use in regular expressions, otherwise - empty string.
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    private function getDocrootRegexpPrefix(): string
+    {
+        return $this->getDocrootName() ? $this->getDocrootName() . '\/' : '';
+    }
+
+    /**
+     * Returns the list of regular expressions to match Drupal files paths and sites/@/settings.@.php files.
+     *
+     * @return array
+     *
+     * @throws \Exception
+     */
+    private function getDrupalExcludes(): array
+    {
+        return [
+            '#^' . $this->getDocrootRegexpPrefix() . 'sites\/.+\/files$#',
+            '#^' . $this->getDocrootRegexpPrefix() . 'sites\/.+\/settings\..+\.php$#',
+        ];
+    }
+
+    /**
+     * Returns the list of regular expressions to match Drupal contrib projects paths (modules, themes and profiles).
+     *
+     * @return array
+     *
+     * @throws \Exception
+     */
+    private function getWebDocrootDrupalExcludes(): array
```

Comment:
I've been thinking about this for a while, but I wasn't sure what to advise that would be better. There are two problems with this technique:

- Sites without a relocated docroot could still be Composer managed
- Sites might place Composer dependencies in locations other than the standard place

I discovered that `composer info --path --format=json` will return the installation paths for all packages placed by Composer, so we could use that here regardless of docroot value to find the locations to exclude from the archive.

Reply:
@greg-1-anderson ok, will consider refactoring this